File: 2024 July/linked list of strings forms a palindrome/main.cpp
// Link: https://www.geeksforgeeks.org/problems/linked-list-of-strings-forms-a-palindrome/1

#include <iostream>
#include <vector>
using namespace std;

struct Node
{
    string data;
    Node *next;

    Node(string x)
    {
        data = x;
        next = NULL;
    }
};

Node *newNode(string str)
{
    Node *new_node = new Node(str);
    new_node->next = NULL;
    return new_node;
}

void print(Node *root)
{
    Node *temp = root;

    while (temp != NULL)
    {
        cout << temp->data << endl;
        temp = temp->next;
    }
}

void createLinkedList(Node *&head, vector<string> elements)
{
    if (elements.size() == 0)
    {
        head = NULL;
        return;
    }

    head = newNode(elements[0]);
    Node *tail = head;

    for (int i = 1; i < elements.size(); i++)
    {
        tail->next = newNode(elements[i]);
        tail = tail->next;
    }
}

class Solution
{
public:
    bool compute(Node *head)
    {
        string str = "";
        Node *temp = head;
        while (temp)
        {
            str = str + temp->data;
            temp = temp->next;
        }
        int l = 0, r = str.size() - 1;
        while (l < r)
        {
            if (str[l] != str[r])
            {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
};

int main()
{
    Solution obj;
    vector<string> elements = {"a", "bc", "d", "dcb", "a"};
    Node *head;
    createLinkedList(head, elements);
    cout << obj.compute(head) << endl;

    elements = {"a", "bc", "d", "ba"};
    createLinkedList(head, elements);
    cout << obj.compute(head) << endl;
    return 0;
}

File: 2024 July/linked list of strings forms a palindrome/main.py
# Link: https://www.geeksforgeeks.org/problems/linked-list-of-strings-forms-a-palindrome/1


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Llist:
    def __init__(self):
        self.head = None

    def insert(self, data, tail):
        node = Node(data)

        if not self.head:
            self.head = node
            return node

        tail.next = node
        return node


def createLinkedList(elements):
    llist = Llist()
    tail = None
    for element in elements:
        tail = llist.insert(element, tail)
    return llist.head


class Solution:
    def compute(self, head):
        string = ""
        temp = head
        while temp:
            string += temp.data
            temp = temp.next

        l, r = 0, len(string) - 1
        while l < r:
            if string[l] != string[r]:
                return False
            l += 1
            r -= 1

        return True


if __name__ == "__main__":
    obj = Solution()
    elements = ["a", "bc", "d", "dcb", "a"]
    head = createLinkedList(elements)
    print(obj.compute(head))

    elements = ["a", "bc", "d", "ba"]
    head = createLinkedList(elements)
    print(obj.compute(head))
